app/api/items/ask_hint.go

```diff
@@ -0,0 +1,244 @@
+package items
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"unsafe"
+
+	"github.com/go-chi/render"
+	"github.com/jinzhu/gorm"
+	"gopkg.in/jose.v1/crypto"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/logging"
+	"github.com/France-ioi/AlgoreaBackend/app/payloads"
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+	"github.com/France-ioi/AlgoreaBackend/app/token"
+)
+
+func (srv *Service) askHint(w http.ResponseWriter, r *http.Request) service.APIError {
+	requestData := AskHintRequest{store: srv.Store, publicKey: srv.TokenConfig.PublicKey}
+
+	var err error
+	if err = render.Bind(r, &requestData); err != nil {
+		return service.ErrInvalidRequest(err)
+	}
+
+	user := srv.GetUser(r)
+	apiError := service.NoError
+	if apiError = checkAskHintUsers(user, &requestData); apiError != service.NoError {
+		return apiError
+	}
+
+	err = srv.Store.InTransaction(func(store *database.DataStore) error {
+		var hasAccess bool
+		var reason error
+		hasAccess, reason, err = store.Items().CheckSubmissionRights(requestData.TaskToken.Converted.LocalItemID, user)
+		service.MustNotBeError(err)
+
+		if !hasAccess {
+			apiError = service.ErrForbidden(reason)
+			return nil // commit! (CheckSubmissionRights() changes the DB sometimes)
+		}
+
+		userItemStore := store.UserItems()
+		err = userItemStore.CreateIfMissing(user.UserID, requestData.TaskToken.Converted.LocalItemID)
+		service.MustNotBeError(err)
+
+		// Get the previous hints requested JSON data
+		var hintsRequestedParsed []payloads.Anything
+		hintsRequestedParsed, err = queryAndParsePreviouslyRequestedHints(&requestData, store, user, r)
+		if err == gorm.ErrRecordNotFound {
+			apiError = service.ErrNotFound(errors.New("can't find previously requested hints info"))
+			return nil // commit
+		}
+		service.MustNotBeError(err)
+
+		// Add the new requested hint to the list if it's not in the list yet
+		hintsRequestedParsed = addHintToListIfNeeded(hintsRequestedParsed, requestData.HintToken.AskedHint)
+
+		var hintsRequestedNew []byte
+		hintsRequestedNew, err = json.Marshal(hintsRequestedParsed)
+		service.MustNotBeError(err)
+		hintsRequestedNewString := string(hintsRequestedNew)
+		requestData.TaskToken.HintsRequested = &hintsRequestedNewString
+		hintsGivenCountString := strconv.Itoa(len(hintsRequestedParsed))
+		requestData.TaskToken.HintsGivenCount = &hintsGivenCountString
+
+		columnsToUpdate := map[string]interface{}{
+			"sHintsRequested":            hintsRequestedNew,
+			"nbHintsCached":              len(hintsRequestedParsed),
```

Comment:
Should we allow a decrease of this number? 🤔 ... one may try to find a flaw in the hint storage to try to reset its counter... shouldn't the counter (which is already an information duplication) be a barrier to this by rejecting a decrease? 
An alternative could be not to count the size of the new list but just count the number of elements added in the list and add this value... meaning that if the JSON cannot be parsed as so it is resetted, the counter does not come back to 0 but continue to increase with the number of hints added. (don't know if it is clear)
The alternative has the advantage not to block everything if something goes wrong, which is maybe better, but it has to be well explained.

Reply:
The root of the problem is that a user can reset sHintsRequested which we should prohibit I suppose.

Reply:
How can he?

Reply:
What if a user saves the oldest task token, then requests some hints, then requests a hint again using the oldest task token and the first hint token?

Reply:
That's why I put the comment below the spec table "Note (dle): Will require full scenarios for checking that after a hint request, there is no way to get a score ignoring there was a hint given." 
Actually, this should not be possible, and that's why the backend saves the history of requested hints. You should just add the hint in the list, not dropping the missing hints... but that's already what you do, don't you? 🤨

Reply:
Yes, that's exactly what I do :)